### PR TITLE
Use environment variables in Dockerfiles entrypoint

### PR DESCRIPTION
**Pull request by ferferga**
Would be also really cool if we include a reference to them in Docker Hub, so people know they exist.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -51,10 +51,12 @@ ENV DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1
 ENV LC_ALL en_US.UTF-8
 ENV LANG en_US.UTF-8
 ENV LANGUAGE en_US:en
+ENV JELLYFIN_DATA_DIR="/config" \
+    JELLYFIN_CACHE_DIR="/cache"
```
Comment (dkanada) [inline Dockerfile:55]: I think these should be separate lines.

Comment (ferferga) [reply Dockerfile]: @dkanada Done!

**Comment by ferferga** [comment]
As pointed out by @Bond-009 in Matrix, environment variables are already used by the server's code at https://github.com/jellyfin/jellyfin/blob/master/Jellyfin.Server/Program.cs#L382 file. This makes the entrypoint's arguments useless. 

Now, we're declaring all the possible environment variables that jellyfin supports, making our containers extremely configurable. Although @dkanada pointed out before that the envs should be in separate lines (I suppose for consistency between the old ones), I think that we shouldn't stick to that now that we're adding more than two variables **because we're creating usless layers and increasing building time**.

I built and tested the x86/amd64 dockerfile and everything is working fine for fresh containers and already existing installations.

**Comment by ferferga** [comment]
As pointed out by @Bond-009 in Matrix, environment variables are already used by the server's code at https://github.com/jellyfin/jellyfin/blob/master/Jellyfin.Server/Program.cs#L382 file. This makes the entrypoint's arguments useless. 

Now, we're declaring all the possible environment variables that jellyfin supports, making our containers extremely configurable. Although @dkanada pointed out before that the envs should be in separate lines (I suppose for consistency between the old ones), I think that we shouldn't stick to that now that we're adding more than two variables **because we're creating usless layers and increasing building time**.

As entrypoints doesn't have env vars initialised, I needed to make a custom one. This allow for easier entrypoint overriding and opens the possibility to run specific scripts during the first boot-up of the image (for configuring other parameters that need to go into config files, for example). This is also a nice step towards making our containers run as non-root (something I plan to work on the future), removing a privilege scalation attack vector we have.

I built and tested the x86/amd64 dockerfile and everything is working fine for fresh containers and already existing installations.

**Comment by nvllsvm** [comment]
I largely disagree with these changes - they add a bunch of complexity for no real gain. 

What problem are you trying to solve?

**Comment by ferferga** [comment]
@nvllsvm I want to make a copy of my current library and data dirs to other locations and make the mountpoints there.

Docker forces you to put the data under /config, but after reusing my production library this breaks whatever testing I can do (because Jellyfin expects data in /var/lib/jellyfin).

Same for web. Now that we're splitting packages, what if I want to test a tagged old web container with a latest nightly server? I might want to place it under other path.

Same for ffmpeg.

Same for non-root. I think it's a must thing we should do but we don't and it might require some sort of configuration at boot time, which can be done with the splitted entrypoint.

It also allows you to easily change what you want to run by appending "/bin/bash" at the end of the run command, instead of forcing Jellyfin and the use of --entrypoint argument.

I know this adds more complexity, but it's how Docker containers should be built imho. I agree with you many options will be barely used, but containers should give as many options possible to get this quick up and running. We still need arguments for setting the default user, the default password, the baseurl, logging level, etc. I just took the inspiration in how most maintained packages in DockerHub and, largely, Bitnami images, are packaged.

**Comment by EraYaN** [comment]
We should probably rename the `entrypoint.sh` to `docker-entrypoint.py` since basically everyone uses that name. And run.sh should maybe be merged into that script as the "default" action.

**Comment by nvllsvm** [comment]
> @nvllsvm I want to make a copy of my current library and data dirs to other locations and make the mountpoints there.
> 
> Docker forces you to put the data under /config, but after reusing my production library this breaks whatever testing I can do (because Jellyfin expects data in /var/lib/jellyfin).

Just mount your local directory to `/config`:
```
docker run -v /local:/config
```


> 
> Same for web. Now that we're splitting packages, what if I want to test a tagged old web container with a latest nightly server? I might want to place it under other path.

No reason to make the container configurable here - just mount your local directory with standard volume syntax
```
docker run -v /local:/jellyfin/jellyfin-web
```

> Same for ffmpeg.

If you're at the point where you need to customize ffmpeg in the Docker container, you should consider making your own Dockerfile that inherits from the jellyfin container. There's too many potential problems with library versions that makes ffmpeg a hard thing to simply drop-in-place.


> We still need arguments for setting the default user,

Docker already provides a mechanism to set UID and GID - the `--user` and `--group-add` arguments to Docker run.

> the default password,

Why?

> the baseurl, logging level,

IMO - these are not exclusive to Docker and can be a part of Jellyfin itself.

**Comment by joshuaboniface** [comment]
So, it's worth noting that once the build changes go through (#3386), these Docker images will be for development only, with the "real" production Dockerfiles under https://github.com/jellyfin/jellyfin-metapackages. Thus I sort of agree that more configurability would be preferable, though all of @nvllsvm's points are perfectly valid alternative solutions to this.

**Comment by joshuaboniface** [comment]
Looking this over a few times, I'm not certain I agree that this is more complex @nvllsvm - it's replacing a bunch of hardcoded defaults with envvars which would be overridable by users. I think this has a lot of benefits and isn't any more complex, in fact it actually reduces complexity in my view since all these options are now in one place at the top of the dockerfile. 

That said, I'm not the biggest fan of two layers of entrypoint. Just having `ENTRYPOINT ["./jellyfin/jellyfin"]` is enough here, since all the options would then be set by the envvars. Doing `exec $@` is bad form and totally unnecessary.

**Comment by joshuaboniface** [comment]
Based on our Matrix chats, this would be the "golden" setup for this, using the Metaimage version from jellyfin/jellyfin-metapackages - same changes could be ported into this PR for the development images:

```
# By default build for stable; unstable must set this explicitly
ARG TARGET_RELEASE=stable

FROM jellyfin/jellyfin-server:${TARGET_RELEASE}-amd64 as server
FROM jellyfin/jellyfin-web:${TARGET_RELEASE} as web 
FROM debian:buster-slim

# Default environment variables for the Jellyfin invocation
ENV DOTNET_SYSTEM_GLOBALIZATION_INVARIANT=1 \
    LC_ALL=en_US.UTF-8 \
    LANG=en_US.UTF-8 \
    LANGUAGE=en_US:en \
    NVIDIA_DRIVER_CAPABILITIES="compute,video,utility" \
    JELLYFIN_MEDIA_DIR="/media" \
    JELLYFIN_DATA_DIR="/config" \
    JELLYFIN_CACHE_DIR="/cache" \
    JELLYFIN_CONFIG_DIR="${JELLYFIN_DATA_DIR}/config" \
    JELLYFIN_LOG_DIR="${JELLYFIN_DATA_DIR}/log" \
    JELLYFIN_WEB_DIR="/jellyfin/jellyfin-web" \
    JELLYFIN_FFMPEG="/usr/lib/jellyfin-ffmpeg/ffmpeg"

# Install dependencies:
#   mesa-va-drivers: needed for AMD VAAPI
RUN apt-get update \
 && apt-get install --no-install-recommends --no-install-suggests -y ca-certificates gnupg wget apt-transport-https \
 && wget -O - https://repo.jellyfin.org/jellyfin_team.gpg.key | apt-key add - \
 && echo "deb [arch=amd64] https://repo.jellyfin.org/debian buster main" > /etc/apt/sources.list.d/jellyfin.list \
 && cat /etc/apt/sources.list.d/jellyfin.list \
 && apt-get update \
 && apt-get install --no-install-recommends --no-install-suggests -y mesa-va-drivers jellyfin-ffmpeg openssl locales libfontconfig1 libfreetype6 \
 && apt-get remove gnupg wget apt-transport-https -y \
 && apt-get clean autoclean -y \
 && apt-get autoremove -y \
 && rm -rf /var/lib/apt/lists/* \
 && sed -i -e 's/# en_US.UTF-8 UTF-8/en_US.UTF-8 UTF-8/' /etc/locale.gen && locale-gen \
 && mkdir -p ${JELLYFIN_MEDIA_DIR} ${JELLYFIN_DATA_DIR} ${JELLYFIN_CACHE_DIR} \
 && chmod 755 ${JELLYFIN_MEDIA_DIR} ${JELLYFIN_DATA_DIR} ${JELLYFIN_CACHE_DIR}                                                                                                                                                                 

COPY --from=server /jellyfin /jellyfin
COPY --from=web /jellyfin-web /jellyfin/jellyfin-web

EXPOSE 8096
VOLUME ${JELLYFIN_MEDIA_DIR} ${JELLYFIN_DATA_DIR} ${JELLYFIN_CACHE_DIR}
ENTRYPOINT [ "/jellyfin/jellyfin" ]
```

Would just need to be tested.

**Comment by ferferga** [comment]
@nvllsvm what you suggest will make the server freak out with existing bare metal databases because the entrypoint. I might have done something wrong, but I couldn't get them to work other way without overriding the entrypoint (removing the arguments) and using the env vars exclusively.

Also, most of those settings are neither useful for me. But they might be for some people. 

The default user one, for example, appeared the other day in Matrix.

**Comment by joshuaboniface** [comment]
Based on our discussions in Matrix today, I think we've come to the following conclusions:

1. The envvars is a go. We all like the idea. My PoC above can be a solid example for porting this back to the main repo's dev Dockerfiles, and will be accepted by me at least.

2. The entrpoint/scripts stuff is a pass for now. It's a pain in the butt for understanding, even if it is what "everyone else does", and it might break things. So let's leave the entrypoing to be the main binary as in the PoC.

I've done some more testing with this config, and I think it does everything you would want using combinations of `--user`, `--volume`, and setting the envvars on the path:

```
joshua@test ~ $ sudo docker run -e JELLYFIN_DATA_DIR=/var/lib/jellyfin -e JELLYFIN_LOG_DIR=/var/lib/jellyfin/log -e JELLYFIN_CONFIG_DIR=/var/lib/jellyfin/config -e JELLYFIN_CACHE_DIR=/var/lib/jellyfin/cache --volume /home/jellyfin:/var/lib/jellyfin --user $( id jellyfin -u ) jellyfin:testing-amd64
[01:48:00] [INF] [1] Main: Jellyfin version: 10.6.0
[01:48:00] [INF] [1] Main: Environment Variables: ["[JELLYFIN_LOG_DIR, /var/lib/jellyfin/log]", "[JELLYFIN_FFMPEG, /usr/lib/jellyfin-ffmpeg/ffmpeg]", "[DOTNET_SYSTEM_GLOBALIZATION_INVARIANT, 1]", "[JELLYFIN_CACHE_DIR, /var/lib/jellyfin/cache]", "[JELLYFIN_WEB_DIR, /jellyfin/jellyfin-web]", "[JELLYFIN_CONFIG_DIR, /var/lib/jellyfin/config]", "[JELLYFIN_DATA_DIR, /var/lib/jellyfin]", "[JELLYFIN_MEDIA_DIR, /media]"]
[01:48:00] [INF] [1] Main: Arguments: ["/jellyfin/jellyfin.dll"]
[01:48:00] [INF] [1] Main: Operating system: Linux
[01:48:00] [INF] [1] Main: Architecture: X64
[01:48:00] [INF] [1] Main: 64-Bit Process: True
[01:48:00] [INF] [1] Main: User Interactive: True
[01:48:00] [INF] [1] Main: Processor count: 3
[01:48:00] [INF] [1] Main: Program data path: /var/lib/jellyfin
[01:48:00] [INF] [1] Main: Web resources path: /jellyfin/jellyfin-web
[01:48:00] [INF] [1] Main: Application directory: /jellyfin/
[01:48:00] [INF] [1] Emby.Server.Implementations.AppBase.BaseConfigurationManager: Setting cache path: /var/lib/jellyfin/cache
[01:48:00] [INF] [1] Emby.Server.Implementations.ApplicationHost: Loading assemblies
[01:48:01] [WRN] [1] Microsoft.EntityFrameworkCore.Model.Validation: The entity type 'AccessSchedule' is configured to use schema 'jellyfin'. SQLite does not support schemas. This configuration will be ignored by the SQLite provider.
[01:48:01] [WRN] [1] Microsoft.EntityFrameworkCore.Model.Validation: The entity type 'ActivityLog' is configured to use schema 'jellyfin'. SQLite does not support schemas. This configuration will be ignored by the SQLite provider.
[01:48:01] [WRN] [1] Microsoft.EntityFrameworkCore.Model.Validation: The entity type 'ImageInfo' is configured to use schema 'jellyfin'. SQLite does not support schemas. This configuration will be ignored by the SQLite provider.
[01:48:01] [WRN] [1] Microsoft.EntityFrameworkCore.Model.Validation: The entity type 'Permission' is configured to use schema 'jellyfin'. SQLite does not support schemas. This configuration will be ignored by the SQLite provider.
[01:48:01] [WRN] [1] Microsoft.EntityFrameworkCore.Model.Validation: The entity type 'Preference' is configured to use schema 'jellyfin'. SQLite does not support schemas. This configuration will be ignored by the SQLite provider.
[01:48:01] [WRN] [1] Microsoft.EntityFrameworkCore.Model.Validation: The entity type 'User' is configured to use schema 'jellyfin'. SQLite does not support schemas. This configuration will be ignored by the SQLite provider.
[01:48:01] [INF] [1] Main: Kestrel listening on all interfaces
[01:48:01] [WRN] [1] Microsoft.AspNetCore.DataProtection.Repositories.EphemeralXmlRepository: Using an in-memory repository. Keys will not be persisted to storage.
[01:48:01] [WRN] [1] Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager: Neither user profile nor HKLM registry available. Using an ephemeral key repository. Protected data will be unavailable when application exits.
[01:48:01] [WRN] [1] Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager: No XML encryptor configured. Key {502403b1-f9a0-4eee-a449-94383a3c1c3b} may be persisted to storage in unencrypted form.
[01:48:01] [INF] [1] Emby.Server.Implementations.ApplicationHost: Running startup tasks
[01:48:01] [INF] [1] Emby.Server.Implementations.ScheduledTasks.TaskManager: Daily trigger for Extract Chapter Images set to fire at 06/23/2020 02:00, which is 0:11:58.0663947 from now.
[01:48:02] [INF] [1] MediaBrowser.MediaEncoding.Encoder.MediaEncoder: FFmpeg: NotFound: 
[01:48:02] [INF] [1] Emby.Server.Implementations.ApplicationHost: ServerId: 0cb82a9381fc4db1a64fc5d54d534a02
[01:48:02] [INF] [10] Emby.Dlna.Main.DlnaEntryPoint: Registering publisher for urn:schemas-upnp-org:device:MediaServer:1 on 172.17.0.2
[01:48:02] [INF] [10] Emby.Server.Implementations.ApplicationHost: Executed all pre-startup entry points in 0:00:00.3000648
[01:48:02] [INF] [10] Emby.Server.Implementations.ApplicationHost: Core startup complete
[01:48:02] [INF] [10] Emby.Server.Implementations.ApplicationHost: Executed all post-startup entry points in 0:00:00.1271105
[01:48:02] [INF] [10] Main: Startup complete 0:00:02.4204175
[01:48:04] [INF] [11] Emby.Server.Implementations.ScheduledTasks.TaskManager: StartupTrigger fired for task: Update Plugins
[01:48:04] [INF] [11] Emby.Server.Implementations.ScheduledTasks.TaskManager: Queueing task PluginUpdateTask
[01:48:04] [INF] [6] Emby.Server.Implementations.ScheduledTasks.TaskManager: Executing Update Plugins
[01:48:05] [INF] [6] Emby.Server.Implementations.ScheduledTasks.TaskManager: Update Plugins Completed after 0 minute(s) and 0 seconds
[01:48:05] [INF] [6] Emby.Server.Implementations.ScheduledTasks.TaskManager: ExecuteQueuedTasks
```

**Comment by ferferga** [comment]
Closing this as we're now doing everything in jellyfin/jellyfin-metapackages#1 instead

**Comment by joshuaboniface** [comment]
Oh, we should do this one too @ferferga since this image will still be used for testing purposes, which is one of the best uses for these changes.

**Pull request closed without merging**